lib/daru/io/exporters/avro.rb

```diff
@@ -0,0 +1,74 @@
+require 'daru/io/exporters/base'
+
+module Daru
+  module IO
+    module Exporters
+      class Avro < Base
+        Daru::DataFrame.register_io_module :to_avro, self
+
+        # Exports +Daru::DataFrame+ to an Avro file.
+        #
+        # @param dataframe [Daru::DataFrame] A dataframe to export
+        # @param path [String] Path of Avro file where the dataframe is to be saved
+        # @param schema [Hash] A schema hash containing keys such as +:type+, +:name+
+        #   and +:fields+
+        #
+        # @example Writing to an Avro file
+        #   schema = {
+        #     "type" => "record",
+        #     "name" => "User",
+        #     "fields" => [
+        #       {"name" => "name", "type" => "string"},
+        #       {"name" => "points", "type" => "int"},
+        #       {"name"=> "winner", "type"=> "boolean", "default"=> "false"}
+        #     ]
+        #   }
+        #
+        #   df = Daru::DataFrame.new(
+        #     [
+        #       {"name"=> "Dany", "points"=> 100, "winner"=> true},
+        #       {"name"=> "Jon", "points"=> 100, "winner"=> true},
+        #       {"name"=> "Tyrion", "points"=> 100, "winner"=> true}
+        #     ]
+        #   )
+        #
+        #   #=> #<Daru::DataFrame(3x3)>
+        #   #        name points winner
+        #   #    0   Dany    100   true
+        #   #    1    Jon    100   true
+        #   #    2 Tyrion    100   true
+        #
+        #   Daru::IO::Exporters::Avro.new(df, "azorahai.avro", schema).call
+        def initialize(dataframe, path, schema=nil)
+          optional_gem 'avro'
+          require 'json'
+
+          super(dataframe)
+          @path   = path
+          @schema = process_schema(schema)
+        end
+
+        def call
+          @schema_obj = ::Avro::Schema.parse(@schema)
+          @writer     = ::Avro::IO::DatumWriter.new(@schema_obj)
+          @buffer     = StringIO.new
+          @writer     = ::Avro::DataFile::Writer.new(@buffer, @writer, @schema_obj)
+          @dataframe.each_row { |row| @writer << row.to_h }
+          @writer.close
+
+          File.open(@path, 'w') { |file| file.write(@buffer.string) }
+        end
+
+        private
+
+        def process_schema(schema)
+          case schema
+          when String then schema
+          when Hash   then schema.to_json
+          else raise ArgumentError, 'Invalid Avro Schema provided.'
```

Comment:
Though raising an `ArgumentError` follows KISS, would having an intelligent guessing method here be useful for users?